widget/list.go

```diff
@@ -0,0 +1,405 @@
+)
+
+// List is a widget that pools list items for performance and
+// lays the items out in a vertical direction inside of a scroller.
+// List requires that all items are the same size
+type List struct {
+	BaseWidget
+
+	Length         func() int
+	CreateItem     func() fyne.CanvasObject
+	UpdateItem     func(index int, item fyne.CanvasObject)
+	OnItemSelected func(index int, item fyne.CanvasObject)
+
+	offsetY         int
+	previousOffsetY int
+	selectedItem    *listItem
+	selectedIndex   int
+	itemMin         fyne.Size
+}
+
+// NewList creates and returns a list widget for displaying items in
+// a vertical layout with scrolling and caching for performance
+func NewList(length func() int, createItem func() fyne.CanvasObject, updateItem func(index int, item fyne.CanvasObject)) *List {
+	list := &List{BaseWidget: BaseWidget{}, Length: length, CreateItem: createItem, UpdateItem: updateItem, selectedIndex: -1}
+	list.ExtendBaseWidget(list)
+	return list
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (l *List) CreateRenderer() fyne.WidgetRenderer {
+	l.ExtendBaseWidget(l)
+
+	if l.itemMin.IsZero() && l.CreateItem != nil {
+		l.itemMin = newListItem(l.CreateItem(), nil).MinSize()
+	}
+	layout := fyne.NewContainerWithLayout(newListLayout(l))
+	layout.Resize(layout.MinSize())
+
+	scroller := NewVScrollContainer(layout)
+
+	scroller.onOffsetChanged = func() {
+		if l.offsetY == scroller.Offset.Y {
+			return
+		}
+		l.offsetY = scroller.Offset.Y
+		l.BaseWidget.Refresh()
+	}
+
+	objects := []fyne.CanvasObject{scroller}
+	return &listRenderer{BaseRenderer: widget.NewBaseRenderer(objects), list: l, scroller: scroller, layout: layout}
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (l *List) MinSize() fyne.Size {
+	l.ExtendBaseWidget(l)
+
+	return l.BaseWidget.MinSize()
+}
+
+type listRenderer struct {
+	widget.BaseRenderer
+	list             *List
+	scroller         *ScrollContainer
+	layout           *fyne.Container
+	itemPool         *listPool
+	children         []fyne.CanvasObject
+	visibleItemCount int
+	firstItemIndex   int
+	lastItemIndex    int
+	size             fyne.Size
+}
+
+func (l *listRenderer) Layout(size fyne.Size) {
+	if size != l.size {
+		if size.Width != l.size.Width {
+			for _, child := range l.children {
+				child.Resize(fyne.NewSize(size.Width, l.list.itemMin.Height))
+			}
+		}
+		l.scroller.Resize(size)
+		l.size = size
+	}
+	if l.list.Length() == 0 {
+		return
+	}
+	if l.itemPool == nil {
+		l.itemPool = &listPool{}
+	}
+	offsetChange := l.list.previousOffsetY - l.list.offsetY
+
+	if offsetChange < 0 {
+		// Scrolling Down
+		l.scrollDown(offsetChange)
+		return
+
+	} else if offsetChange > 0 {
+		// Scrolling Up
+		l.scrollUp(offsetChange)
+		return
+	}
+
+	// Relayout What Is Visible - no scroll change - initial layout or possibly from a resize
+	for _, child := range l.children {
+		l.itemPool.Release(child)
+	}
+	l.children = nil
```

Comment:
Sorry I am confused by this code block.
the comment above says no scroll change so I don't quite get why it is reset. 
However, even if the reset is needed then l.children is set to nil, and then you check if it is longer than the visible items - but that will never be true will it?

Reply:
This is fixed now.  No need to reset things.  Just ensure the proper count of items and update them.